### DISCUSSION TITLE
Sort by semantic version order / Custom SQLite functions

### DISCUSSION THREAD

**Discussion by DigNZ** [pocketbase/pocketbase#2940] [Q&A] [answered]
Semantic versioning seems to be a tricky thing to sort by in SQL (1.11.0, 1.12.3 etc.,). I believe this is also a natural sort instead of numeric or alphabetic sorting.

I have collection that has details of devices and the current version on the app they are running - I would like this list in order of app version, also a view table of the distinct app versions in use. 

One of the reasons for the abstraction to the view table (and not being able to do it in the client) is that I need to be able to request the latest version across all devices without exposing the devices to other accounts. I hope that makes sense.

Anyway - I could achieve this with a custom SQLite function - I'm using pocketbase as a framework. What is the best way to register a custom function into SQLite so I can use it in a view definition in pocketbase or are there other ways I could make a SQL query to order by semantic version.

**Comment by ganigeorgiev** [comment, marked as answer]
There is no easy way to register a custom SQLite scalar function because both the CGO driver and the pure Go SQLite port have different interface for that, and additionally the performance is not that great (_at least based on my earlier match-all implementation with the pure Go version_).

If the version strings shares the same format and always have 3 parts, then you can combine [`substr` and `instr` built-in functions](https://www.sqlite.org/lang_corefunc.html) and order by each part of the version, something like:

```sql
select
   id,
   version,
    substr(version, 1, instr(version, '.') - 1) as v1,
    substr(substr(version, instr(version, '.') + 1), 1, instr(substr(version, instr(version, '.') + 1), '.') - 1) as v2,
    substr(substr(version, instr(version, '.') + 1), instr(substr(version, instr(version, '.') + 1), '.') + 1) as v3
from devices
order by
    cast(v1 as real) asc,
    cast(v2 as real) asc,
    cast(v3 as real) asc
```
_(you may have to adjust the above as I only briefly tested it and may not work with all cases)_

**Comment by DigNZ** [comment]
Thanks

That answers both questions - I've used custom sqlite functions in the past (without pocketbase) with some success but they are obviously a little slower. Your example SQL works within limitations of the parts of the version numbers all being numeric and only being 3 of them but we can work with that.
